### PR TITLE
[CARBONDATA-2224][File Level Reader Support] External File level reader support

### PR DESCRIPTION
**Pull request by sounakr**
File level reader reads any carbondata file placed in any external file path. The reading can be done through 3 methods.
a) Reading as a datasource from Spark. CarbonFileLevelFormat.scala is used in this case to read the file. To create a spark datasource external table
" CREATE TABLE sdkOutputTable USING CarbonDataFileFormat LOCATION '$writerOutputFilePath1'"
For more details please refer the test file org/apache/carbondata/spark/testsuite/createTable/TestCreateTableUsingCarbonFileLevelFormat.scala
file.

b) Reading from spark sql as a external table. CarbonFileinputFormat.java is used for reading the files. The create table syntax for this will be
"CREATE EXTERNAL TABLE sdkOutputTable STORED BY 'carbondatafileformat' LOCATION '$writerOutputFilePath6'"
For more details org/apache/carbondata/spark/testsuite/createTable/TestCarbonFileInputFormatWithExternalCarbonTable.scala.

c) Reading Through Hadoop Map reduce job. Please refer org/apache/carbondata/mapred/TestMapReduceCarbonFileInputFormat.java for more details.

**Limitation** :: This implementation depend on writer SDK file path as following table_name/Fact/Part0/Segment_null.  This reader writer must be independent of static path.
Due to this reader currently won't work with standard partition also. This will be handled in future PRs.

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2964/

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java [hunk 1]
```diff
@@ -826,6 +826,12 @@ public boolean isExternalTable() {
     return external != null && external.equalsIgnoreCase("true");
   }
 
+  public boolean isFileLevelExternalTable() {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: why is this property required?

File: hadoop/src/main/java/org/apache/carbondata/hadoop/util/SchemaReader.java [hunk 2]
```diff
@@ -28,7 +28,8 @@
 import org.apache.carbondata.core.metadata.schema.table.TableInfo;
 import org.apache.carbondata.core.util.CarbonUtil;
 import org.apache.carbondata.core.util.path.CarbonTablePath;
-import org.apache.carbondata.core.util.path.CarbonTablePath;
+
+
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/util/SchemaReader.java]: remove empty line

File: hadoop/src/main/java/org/apache/carbondata/hadoop/util/SchemaReader.java [hunk 3]
```diff
@@ -79,4 +81,19 @@ public static TableInfo getTableInfo(AbsoluteTableIdentifier identifier)
         carbonTableIdentifier.getTableName(),
         identifier.getTablePath());
   }
+
+
+  public static TableInfo inferSchemaForExternalTable(AbsoluteTableIdentifier identifier)
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/util/SchemaReader.java]: Can the input param change to `String tablePath`?

File: store/sdk/src/test/java/org/apache/carbondata/sdk/file/CSVCarbonWriterSuite.java [hunk 4]
```diff
@@ -68,13 +68,12 @@ public void testWriteFilesJsonSchema() throws IOException {
 
   private void writeFilesAndVerify(Schema schema, String path) {
     try {
-      CarbonWriter writer = CarbonWriter.builder()
-          .withSchema(schema)
-          .outputPath(path)
-          .buildWriterForCSVInput();
+      CarbonWriter writer =
+          CarbonWriter.builder().withSchema(schema).outputPath(path).buildWriterForCSVInput();
```
Comment (jackylk) [inline store/sdk/src/test/java/org/apache/carbondata/sdk/file/CSVCarbonWriterSuite.java]: do not modify the code style

File: store/sdk/src/test/java/org/apache/carbondata/sdk/file/CSVCarbonWriterSuite.java [hunk 5]
```diff
@@ -68,13 +68,12 @@ public void testWriteFilesJsonSchema() throws IOException {
...
 
       for (int i = 0; i < 100; i++) {
-        writer.write(new String[]{"robot" + i, String.valueOf(i), String.valueOf((double) i / 2)});
+        writer
+            .write(new String[] { "robot" + i, String.valueOf(i), String.valueOf((double) i / 2) });
```
Comment (jackylk) [inline store/sdk/src/test/java/org/apache/carbondata/sdk/file/CSVCarbonWriterSuite.java]: do not modify it since no change

**Comment by jackylk** [comment]
There are some binary files, please delete them

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4210/

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java (hunk 1 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: **stored by 'carbondatafileformat' is mapped with _filelevelexternal.**
So, In carbonScanRDD, when mapReduce service or hadoop service calls carbonScanRDD.
based on _filelevelexternal, new File level reader [CarbonFileInputFormat] will be called.

External table can be table level (stored by 'carbondata') or file level (stored by 'carbondatafileformat')
**This is used to identify the file level external table.**

**Comment by ajantha-bhat** [comment]
@jackylk : Binary files [carbondata and index files (sdk Writer output)] are intentionally added for test cases of this requirement. Test cases will fail If we remove them.

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3858/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4218/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2973/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3867/

**Pull request closed without merging**